satpy/tests/reader_tests/test_viirs_l2.py

```diff
@@ -0,0 +1,155 @@
+#!/usr/bin/env python
+# -*- coding: utf-8 -*-
+# Copyright (c) 2022 Satpy developers
+#
+# This file is part of satpy.
+#
+# satpy is free software: you can redistribute it and/or modify it under the
+# terms of the GNU General Public License as published by the Free Software
+# Foundation, either version 3 of the License, or (at your option) any later
+# version.
+#
+# satpy is distributed in the hope that it will be useful, but WITHOUT ANY
+# WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR
+# A PARTICULAR PURPOSE.  See the GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License along with
+# satpy.  If not, see <http://www.gnu.org/licenses/>.
+"""Tests for the VIIRS CSPP L2 readers."""
+
+import os
+import shutil
+import tempfile
+
+import numpy as np
+import xarray as xr
+
+from satpy import Scene
+
+CLOUD_MASK_FILE = "JRR-CloudMask_v3r0_npp_s202212070905565_e202212070907207_c202212071932513.nc"
+NUM_COLUMNS = 3200
+NUM_ROWS = 768
+DATASETS = ['Latitude', 'Longitude', 'CloudMask', 'CloudMaskBinary']
+
+
+class JRRCloudMaskFile:
+    """Write a test JRR CloudMask file."""
+
+    def __init__(self):
+        """Initialize the class."""
+        self._dir = tempfile.mkdtemp()
```

Comment:
Would it be possible to use the pytest `tmp_path` fixture so you don't need to take care of the cleanup yourself?

Reply:
In which way? Have the fixture passed to every test function and then pass the fixture to `JRRCloudMaskFile.__init__()`? When would the temp directory removed?

Reply:
I think it'd be possible to use a class-scoped fixture.  In this case the directory would be created before the first test in the class and deleted after the last one.  There is also `tmp_path_factory` which is, I think, always session-scoped.  Personally, I would just pass the (function scoped) fixture (or a fixture using the `tmp_path` fixture) explicitly to every test function needing it.

I don't feel strongly either way though; just a suggestion, no need to do anything as far as I'm concerned if you prefer to handle the `mkdtemp` yourself the way you've done.

Reply:
`tmp_path` has the advantage of taking care of removing the files after it, creating custom path for each tests, etc...

Reply:
also, I would make a fixture that returns a created file using this class and tmp_path.

Reply:
So wrapping the `CRRCloudMaskFile` class in a function that returns the filename of the created file, which has the directory part coming from `tmp_path` fixture, and then wrapping the new function with `@pytest.fixture`?

I'm not sure how that works, and is very confusing and obfuscating in general, but I'll try it tomorrow.

Reply:
Something like this (you shouldn't need the `JRRTolkienCloudmaskFile` class anymore:

```python
@pytest.fixture
def cloud_mask_file(tmp_path):
    path = tmp_path / CLOUD_MASK_FILE
	write_cloudmask_file(path)
	return path


def write_cloudmask_file(path):
        dset = xr.Dataset()
        dset.attrs = _get_global_attrs()
        dset['Latitude'] = _get_lat_arr()
        dset['Longitude'] = _get_lon_arr()
        dset['CloudMask'] = _get_cloud_mask_arr()
        dset['CloudMaskBinary'] = _get_cloud_mask_binary_arr()
        dset.to_netcdf(path, 'w')
```
and then in the test functions:

```python
def test_cloud_mask_read_latitude(cloud_mask_file):
    """Test reading latitude dataset."""
    data = _read_viirs_l2_cloud_mask_nc_data(cloud_mask_file, 'latitude')
    _assert_common(data)
```

I understand that fixtures can appear obfuscating, but I think they help making the testing clearer. The way I look at them is as context. So I create a context with the `@pytest.fixture` decorator, and then I just use the context in the test function. Pytest just takes care of passing the relevant contexts to the test functions.

Reply:
Should've been clearer about what I think is obfuscation in the fixtures: Those fixtures that are not declared in any way in the file. If they are not defined, the shouldn't exist. At the very least they should be explicitly imported, either directly or used via their parent package.

Didn't even know about `conftest.py`, sounds ... scary.

Reply:
> I understand that fixtures can appear obfuscating, but I think they help making the testing clearer.

Well, it depends, see above. Excplicit vs. implicit vs. not-even-implied. 

> The way I look at them is as context. So I create a context with the `@pytest.fixture` decorator, and then I just use the context in the test function. Pytest just takes care of passing the relevant contexts to the test functions.

The context manager in Python is much more clear. It clearly defines a block of code where it has any effect. With a fixture it's not clear where the context starts and where it ends: the function in question, the whole module, even a whole package?

Thanks for the example, I'll check it tomorrow.

Reply:
```python
@pytest.fixture
def cloud_mask_file(tmp_path):
    path = tmp_path / CLOUD_MASK_FILE
    write_cloudmask_file(path)
    yield path
    # any cleanup code goes here
```
Wasn't the point of the `tmp_path` that there wouldn't need to be any additional cleanup?

Reply:
```python
satpy/tests/reader_tests/test_viirs_l2.py:161: in _read_viirs_l2_cloud_mask_nc_data
    scn = Scene(reader="viirs_l2_cloud_mask_nc", filenames=[fname])
satpy/scene.py:131: in __init__
    filenames = convert_remote_files_to_fsspec(filenames, storage_options)
satpy/utils.py:610: in convert_remote_files_to_fsspec
    return _check_file_protocols(filenames, storage_options)
satpy/utils.py:622: in _check_file_protocols
    local_files, remote_files, fs_files = _sort_files_to_local_remote_and_fsfiles(filenames)
satpy/utils.py:639: in _sort_files_to_local_remote_and_fsfiles
    elif urlparse(f).scheme in ('', 'file') or "\\" in f:
../../../../mambaforge/envs/py310/lib/python3.10/urllib/parse.py:392: in urlparse
    url, scheme, _coerce_result = _coerce_args(url, scheme)
../../../../mambaforge/envs/py310/lib/python3.10/urllib/parse.py:128: in _coerce_args
    return _decode_args(args) + (_encode_result,)
../../../../mambaforge/envs/py310/lib/python3.10/urllib/parse.py:112: in _decode_args
    return tuple(x.decode(encoding, errors) if x else '' for x in args)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

.0 = <tuple_iterator object at 0x7fab3cfb0370>

>   return tuple(x.decode(encoding, errors) if x else '' for x in args)
E   AttributeError: 'PosixPath' object has no attribute 'decode'
```

I thought this was alredy fixed. And if instead of dividing the `tmp_path` with `CLOUD_MASK_FILE` I use the traditional `os.path.join()`, the temporary files are not cleaned.

Reply:
Even when using the division version the temporary files are left behind. I'll stick to the original implementation that does clean the files.

Reply:
tmp_path creates a temporary path that is specific to a given test run. The files are left behind after the test run, so that one can examin them if needed. However, subsequent runs of the tests will create new paths and discard the old ones, so the space used by the temp files aren't growing. Hence "automatic cleanup" in my mind. I would still recommend it.

Reply:
After running the tests three times:
```
$ ls -1 /tmp/pytest-of-lahtinep/
pytest-0
pytest-1
pytest-2
pytest-current
$ du -ch /tmp/pytest-of-lahtinep/
24M	/tmp/pytest-of-lahtinep/pytest-1/test_cloud_mask_read_cloud_mas0
24M	/tmp/pytest-of-lahtinep/pytest-1/test_cloud_mask_read_latitude0
24M	/tmp/pytest-of-lahtinep/pytest-1/test_cloud_mask_read_longitude0
24M	/tmp/pytest-of-lahtinep/pytest-1/test_cloud_mas_read_binary_clo0
94M	/tmp/pytest-of-lahtinep/pytest-1
24M	/tmp/pytest-of-lahtinep/pytest-0/test_cloud_mask_read_cloud_mas0
24M	/tmp/pytest-of-lahtinep/pytest-0/test_cloud_mask_read_latitude0
24M	/tmp/pytest-of-lahtinep/pytest-0/test_cloud_mask_read_longitude0
24M	/tmp/pytest-of-lahtinep/pytest-0/test_cloud_mas_read_binary_clo0
94M	/tmp/pytest-of-lahtinep/pytest-0
24M	/tmp/pytest-of-lahtinep/pytest-2/test_cloud_mask_read_cloud_mas0
24M	/tmp/pytest-of-lahtinep/pytest-2/test_cloud_mask_read_latitude0
24M	/tmp/pytest-of-lahtinep/pytest-2/test_cloud_mask_read_longitude0
24M	/tmp/pytest-of-lahtinep/pytest-2/test_cloud_mas_read_binary_clo0
94M	/tmp/pytest-of-lahtinep/pytest-2
282M	/tmp/pytest-of-lahtinep/
282M	total
```
No cleanup. At all, as far as I can see.

Reply:
try a couple times more.

Reply:
from https://docs.pytest.org/en/7.1.x/reference/reference.html#pytest.tmpdir.tmp_path :
> By default, a new base temporary directory is created each test session, and old bases are removed after 3 sessions, to aid in debugging. If --basetemp is used then it is cleared each session. See [The default base temporary directory](https://docs.pytest.org/en/7.1.x/how-to/tmp_path.html#base-temporary-directory).

Reply:
Forgot to update. Indeed, on the next run the `0` index directories were deleted and new `3` indexed were created. This could potentially leave quite a lot of stuff behind, so maybe there should be a more strict cleaning. But that's another topic outside of this PR.

Reply:
Oh, and now the tests use `tmp_path` fixture. I also added a note at the top of the file listing the external fixtures that are used. I'll be creating a PR discussing this in the dev documentation, and possibly also adding the notes to other tests having this issue of seemingly undefined fixtures in them.